src/Features/Core/Portable/Completion/CompletionServiceWithProviders.cs

```diff
@@ -447,55 +446,33 @@ private static bool HasAnyItems(CompletionContext cc)
                 foreach (var item in context.Items)
                 {
                     Debug.Assert(item != null);
-                    AddToDisplayMap(item, displayNameToItemsMap);
+                    displayNameToItemsMap.Add(item);
                 }
 
                 // first one wins
                 suggestionModeItem ??= context.SuggestionModeItem;
             }
 
-            if (displayNameToItemsMap.Count == 0)
+            if (displayNameToItemsMap.IsEmpty)
             {
                 return CompletionList.Empty;
             }
 
             // TODO(DustinCa): Revisit performance of this.
-            var totalItems = displayNameToItemsMap.Values.Flatten().ToList();
-            totalItems.Sort();
+            using var _ = ArrayBuilder<CompletionItem>.GetInstance(itemsCount, out var builder);
+            builder.AddRange(displayNameToItemsMap);
+            builder.Sort();
+
+            displayNameToItemsMap.Free();
```

Comment:
so i odn't like these manual frees.  because early returns are possible without hitting this.  If this is just a dictionary, can we just use a pooled dictionary instead?

Reply:
I can move this to a `finally`. We don't use pooled dictionary because we want to make sure we have a dedicated pool (a Dictionary<string, object> isn't uncommon).

Reply:
even with your own type, you can make it disposable so you control free'ing things always without needing try/finally.  or, put another way, the `using` pattern is the nicer looking way of doing try/finally :)

Reply:
Sure